docs/exceptions.md

```diff
@@ -0,0 +1,266 @@
+number, but we have to consider the case that the user enters something else.
+
+In cases when the erroneous scenarios are out of our control, we consider
+these to be regular errors.
+
+### Identifying error class
+
+Can the execution of the error codepath be excluded by code review and static
+verification (more precise types?)
+
+* Yes: it is a programmer mistake
+* No: it is a regular error
+
+## Is the code 100% pure or potentially impure?
+
+### Impure code
+
+The code is considered impure when it's an `IO` action, a function that returns
+an `IO` action, or similar. The code can be _potentially_ impure when it is
+written in an abstract monad that can be instantiated to `IO`.
+
+Definitely impure:
+
+```
+readFile :: FilePath -> IO String
+```
+
+Potentially impure:
+
+```
+lookupThrow :: MonadThrow m => Map k v -> k -> m v
+```
+
+### Pure code
+
+Pure code is one that does not use `IO` operations. For example:
+
+```
+lookupMaybe :: Map k v -> k -> Maybe v
+```
+
+Notice that while it is conceivable that we have a `Map` of `IO` actions, and
+instantiate `lookupMaybe` to `Map k (IO ()) -> k -> Maybe (IO ())`, the
+instantiation does not inspect/use these `IO` actions, so it's not considered
+potentially impure.
+
+### Identifying code purity
+
+Is there an instantiation of type parameters that would mean that there are
+`IO` actions inspected/used in the code?
+
+* Yes: it is potentially impure code
+* No: it is 100% pure code
+
+## Error handling practices
+
+### Pure code, programmer mistakes
+
+Do *not*:
+
+* do this often, as if it's okay
+* use `undefined`
+* use non-exhaustive pattern matching
+
+Do:
+
+* try to use types to avoid the need in the first place
+* comment extensively (invariants and precondition, reasoning)
+* use the `error` function (or `impureThrow` with custom exception)
+* use the `HasCallStack` feature
+
+Before writing code that allows programmer mistakes, consult with colleagues how
+it would be possible to prevent them statically. There might be tricks that
+you're not aware of. Always make this trade-off conciously and responsibly.
+
+(On the other hand, if static guarantees require GADTs or type families, perhaps
+it's better to not overcomplicate code. Use your judgement.)
+
+DISCUSSION: Should we create a synonym `bug = impureThrow` in Universum? This
+would make the intention more clear.
+
+### Pure code, regular errors
+
+Do *not*:
+
+* use `error` or `impureThrow`
+* use `MonadFail`
+* return `Either Text`
+
+Do:
+
+* return `Either ErrorADT`, `Maybe`
+* wrap the underlying (pure!) monad in `ExceptT` or `CatchT`
+* use `MonadError` or `MonadThrow`
+
+Consider parsing: it is pure, but we cannot make assumptions about the input. In
+this case we might want to use `ExceptT ParseError`. Or consider a lookup in a
+`Map`, where we don't know whether the key is present -- in this case we'd like
+to return `Maybe v`. In 100% pure code, use one of these ways to handle errors:
```

Comment:
A `Left ParseError` from parsing or a `Nothing` from a map lookup is not an error / exception, it's completely normal. If I get a parse error, nothing has gone wrong, the parser simply decided that the input was not in the language. When I get `Nothing` from a map lookup, nothing has gone wrong, the thing just wasn't in the map... completely normal.

Reply:
That's why this is classified as "regular errors", everything is normal, we just have to handle these cases, which are not the main (desired) flow of events.

Reply:
Right my issue is with the term "error". They're not errors, they are part of the main desired flow of events.

Reply:
What you consider an error depends on your expectations. When you `readFile` and the file does not exist, is it an error? No, we should've expected this. When you allocate a chunk of memory and the heap is full, is this an error? No, we are aware that the events can take this turn. If you think this way, nothing is an error.

Reply:
> When you readFile and the file does not exist, is it an error? ... When you allocate a chunk of memory and the heap is full, is this an error?

Yes and yes. Strictly speaking I'd call them exceptions but let's got hung up on semantics of error vs. exception.

> What you consider an error depends on your expectations

You're absolutely right. When I parse something I expect that the input string may not be recognized. When I lookup a key in a map I expect that it may not be in the map. Nothing went wrong in these cases. But with `readFile`, and with `IO` in general, things may go wrong for reasons outside of the programmer's control (reasons external to the program itself), and there's no obvious one-size-fits-all response to it, so we resort to exception handling.